### ISSUE TITLE
Projecting SMPL body model to the image

### ISSUE THREAD

**Issue by mks0601** [classner/up#17]
Hi,

I'm trying to prjoect SMPL body model to the image using pytorch.
The SMPL body model is obtained from `pose` and `betas` of the `00001_body.pkl` file in UP-3D dataset. Actual mesh model is obtained using [this repo](https://github.com/gulvarol/smplpytorch).

I assume the vertex coordinates are in world coordinate system, so I tried to transform it to the camera coordinate system, and project it to the image coordinate.

First, I added `trans` to the world coordinate mesh coordinates.
`world_coord = world_coord + trans`

Second, I changed the world coordinates to the camera coordinates. I think `rt` and `t` are camera rotation and translation (i.e., extrinsic camera parameters). It seems `rt` is in a form of axis-angle representation, so I converted `rt` to rotation matrix using rodrigues function. The camera coordinates are obtained as follow:
`cam_coord = torch.matmul(rt, (world_coord - t).permute(1,0)).permute(1,0)`.

Third, I projected camera coordinates to the image coordinates. As there is no principal component annotation (i.e., intrinsic camera parameter), I just used half of each image resolution. For example, if the image size is (H,W), then, c = (H/2, W/2). The image coordinates are obtained as follow:
`img_coord_x = cam_coord_x / cam_coord_z * f_x + c_x`.

And the result projected joint coordinates are visualized in this way.
As you can see, it does not align to the image. Could you help me? Am I doing something wrong?

![a](https://user-images.githubusercontent.com/13601723/67058390-9392ce80-f18f-11e9-9f4a-daf1e9ae4327.jpg)

**Comment by classner** [comment]
Hi @mks0601 ,

have you tried comparing your transformations to the one in https://github.com/classner/up/blob/master/3dfit/render.py ? This produces matching renders and you should be able to reproduce everything with Güls code in smplpytorch (unless she's using different coordinate system conventions, etc.).

**Comment by mks0601** [comment]
I read that code, but it seems it renders 3D human model (i.e., SMPL) to 2D image. What I want to get is image-projected (x,y) coordinates and camera coordinate depths of all mesh vertices. And I used perspective projection for that as described in my first issue. Could you tell me what is wrong with my perspective projection formula described above?

**Comment by classner** [comment]
Hi @mks0601 ,

I am not sure where exactly your code breaks, but there are many details that you have to get right (such as the order of operations). Btw., you should be able to get what you want from the code I sent you the link to as well, but not for pytorch. At the very least you should be able to get values for debugging from it and should be able to work backward the projection operations.

**Comment by mks0601** [comment]
I tried to understand that code, but it is wrapped with opendr. In opendr, the definition of ProjectPoint function is not clear for me. I think my question is very easy and simple, so could you just let me know what I've done wrong? It is just a simple image projection from world coordinate system.

```
pose_params = torch.from_numpy(np.array(data['pose'],dtype=np.float32).reshape(1,72)) # smpl theta
shape_params = torch.from_numpy(np.array(data['betas'],dtype=np.float32).reshape(1,10)) # smpl beta
trans = torch.FloatTensor(data['trans']).reshape(1,3) # smpl trans
R, t = rodrigues(torch.FloatTensor(data['rt']).reshape(1,1,3)), torch.FloatTensor(data['t']).reshape(1,1,3) # camera extrinsics
focal = torch.FloatTensor((data['f'], data['f'])).reshape(1,1,2) # camera intrinsic
princpt = torch.FloatTensor((crop_info[1]/2, crop_info[0]/2)).reshape(1,1,2) # camera intrinsic
gt_joint_2d = data['j2d'][:-1,:] # gt_joint_2d: same with smpl + head top (last index). make joint same with SMPL joint
gt_joint_2d = torch.from_numpy(gt_joint_2d).reshape(1,-1,2).float()

 # Forward from the SMPL layer
smpl_vertex, smpl_joint = smpl_layer(pose_params, shape_params, trans)

# world -> camera
smpl_vertex = torch.bmm(R, smpl_vertex.permute(0,2,1)).permute(0,2,1) + t
smpl_joint = torch.bmm(R, smpl_joint.permute(0,2,1)).permute(0,2,1) + t
# camera -> image
smpl_vertex[:,:,:2] = smpl_vertex[:,:,:2] / smpl_vertex[:,:,2:] * focal + princpt
smpl_joint[:,:,:2] = smpl_joint[:,:,:2] / smpl_joint[:,:,2:] * focal + princpt

# for debug
pixel_err = torch.mean(torch.sqrt(torch.sum((smpl_joint[:,:,:2] - gt_joint_2d)**2,2))).numpy()
print('filename: ' + str(filename) + ' pixel error: ' + str(pixel_err))
```

And the result image is 
![0](https://user-images.githubusercontent.com/13601723/68495023-bd24ae80-0292-11ea-87a2-77e73ffc3335.jpg)
where green dots are `gt_joint_2d` and blue dots are `smpl_joint[:,:,:2]`.

**Comment by classner** [comment]
Hi @mks0601,

In this case I'd just step through your script and the one using OpenDR and compare the intermediate results to track down the source of the difference and make sure you're following the same order of transformations.

**Comment by mks0601** [comment]
I tried it.

1. At 217th line of `up/render.py`, it does render.
`renderings = render(model['neutral'], image, cam, steps, segmented, scale)`.

2. At 156th line of `up/render.py` in `render` function, it creates renderer.
```
    rn = _create_renderer(w=int(w * scale),
                          h=int(h * scale),
                          near=4.,
                          far=30.+dist,
                          rt=np.array(cam['rt']),
                          t=np.array(cam['t']),
                          f=np.array([cam['f'], cam['f']]) * scale,
                          # c=np.array(cam['cam_c']),
                          texture=texture)
```.

2-A. At 63th line of `up/render.py` in `_create_renderer` function, it projects SMPL vertices to image plane.
`rn.camera = ProjectPoints(rt=rt, t=t, f=f, c=c, k=k)`.

3. At 170th line of 'up/render.py' in `render` function, it iteratively render mesh from different view.
```
        imtmp = _simple_renderer(rn=rn,
                                 meshes=[mesh],
                                 yrot=light_yrot,
                                 texture=texture)
```.

4. At 131th line of 'up/render.py' in `_simple_renderer` function, it returns `rn.r`. 
`return rn.r`. I need to know what is `r`.

5. At 81th line of ['opendr/camera.py'](https://github.com/mattloper/opendr/blob/master/opendr/camera.py), in `compute_r` function, it returns `self.r_and_derivatives[0].squeeze()`.

6. At 143th line of 'opendr/camera.py' in `r_and_derivatives` function, it calls `cv2.projectPoints`. Here, there are five inputs of this function. `v, self.rt.r, self.t.r, self.camera_mtx, self.k.r`. `rt` and `t` are defined in 156th line of `up/render.py`.
```
self.rt.r: np.array(cam['rt'])
self.t.r: t=np.array(cam['t'])
```
`k` is defined in 56th line of 'up/render.py' as `k = np.zeros(5)`.

`camera_mtx` is obtained from 134th line of 'opendr/camera.py', and 'f' is from 156th line of `up/render.py` `c` is from 55th line of `up/render.py`.
`f=np.array([cam['f'], cam['f']]) * scale,`
 `c = np.array([w, h]) / 2.`
where `w` and `h` are image resolution from 214th line of `up/render.py`.

The result image looks like below. Green dots are `j2d` and blue dots are my projected image points.
![2](https://user-images.githubusercontent.com/13601723/68518340-d77c7d80-02ce-11ea-96a3-7a8d5d01fb03.jpg)

When I use  `c = np.array([crop_info[1], crop_info[0]]) / 2.` like `direct3d/tools/create_dataset.py`, the result image looks like below. Green dots are `j2d` and blue dots are my projected image points. As I defined `c` using `crop_info`, I cropped image using `crop_info[2], [3], [4], [5]' and resized it to `crop_info[0], [1]`.
![2](https://user-images.githubusercontent.com/13601723/68518377-0692ef00-02cf-11ea-9943-070a369adf98.jpg)


Note that I used `cv2.projectPoints` function like opendr. Could you tell me what I've done wrong? and why the result image looks strange when I use  `c = np.array([w, h]) / 2.`?

**Comment by classner** [comment]
Hi @mks0601,

this looks roughly plausible, but unfortunately I don't have the time at hand to support you in-depth.

> I tried it.

It seems like you're missing out on a detail in understanding each renderer and have not executed the code but just looked at it. My recommendation is to execute the code and look at the step-by-step representations to make sure they match.

**Comment by dorianhenning** [comment]
@mks0601 Have you fixed your issue? 
I'm running into the same problem, it seems to me that the joints are in a different coordinate system than the vertices, since the mesh rendering works fine for me, but for all joints and a simple projection it does not work.

would me cool if you found a solution.

**Comment by mks0601** [comment]
I tried to visualize mesh using the same visualization function of the joints (not openDR), but it suffers from the same problem.
I think there is some tricky thing in openDR.
I couldn't fix this problem :(

**Comment by chaowentao** [comment]
it‘s `perspective_projection` function from https://github.com/nkolot/SPIN, it maybe helps you.
```
def perspective_projection(points, rotation, translation, focal_length,
                           camera_center):
    """
    This function computes the perspective projection of a set of points.
    Input:
        points (bs, N, 3): 3D points
        rotation (bs, 3, 3): Camera rotation
        translation (bs, 3): Camera translation
        focal_length (bs,) or scalar: Focal length
        camera_center (bs, 2): Camera center
    """
    batch_size = points.shape[0]
    K = torch.zeros([batch_size, 3, 3], device=points.device)
    K[:, 0, 0] = focal_length
    K[:, 1, 1] = focal_length
    K[:, 2, 2] = 1.0
    K[:, :-1, -1] = camera_center

    # Transform points
    points = torch.einsum("bij,bkj->bki", rotation, points)
    points = points + translation.unsqueeze(1)

    # Apply perspective distortion
    projected_points = points / points[:, :, -1].unsqueeze(-1)

    # Apply camera intrinsics
    projected_points = torch.einsum("bij,bkj->bki", K, projected_points)

    return projected_points[:, :, :-1]

outline_points_2d = perspective_projection(
            pred_vertices[:, list_id, :],
            rotation=torch.eye(3, device=device).unsqueeze(0).expand(
                batch_size, -1, -1),
            translation=camera_translation,
            focal_length=constants.FOCAL_LENGTH,
            camera_center=camera_center,
        )
        # [-h/2, h/2] + h/2 -> [0, h] shape [1,1,2]
        outline_points_2d[:, :, 0] = outline_points_2d[:, :, 0] + width / 2
        outline_points_2d[:, :, 1] = outline_points_2d[:, :, 1] + height / 2
```

**Comment by Walter0807** [comment]
> it‘s `perspective_projection` function from https://github.com/nkolot/SPIN, it maybe helps you.
> 
> ```
> def perspective_projection(points, rotation, translation, focal_length,
>                            camera_center):
>     """
>     This function computes the perspective projection of a set of points.
>     Input:
>         points (bs, N, 3): 3D points
>         rotation (bs, 3, 3): Camera rotation
>         translation (bs, 3): Camera translation
>         focal_length (bs,) or scalar: Focal length
>         camera_center (bs, 2): Camera center
>     """
>     batch_size = points.shape[0]
>     K = torch.zeros([batch_size, 3, 3], device=points.device)
>     K[:, 0, 0] = focal_length
>     K[:, 1, 1] = focal_length
>     K[:, 2, 2] = 1.0
>     K[:, :-1, -1] = camera_center
> 
>     # Transform points
>     points = torch.einsum("bij,bkj->bki", rotation, points)
>     points = points + translation.unsqueeze(1)
> 
>     # Apply perspective distortion
>     projected_points = points / points[:, :, -1].unsqueeze(-1)
> 
>     # Apply camera intrinsics
>     projected_points = torch.einsum("bij,bkj->bki", K, projected_points)
> 
>     return projected_points[:, :, :-1]
> 
> outline_points_2d = perspective_projection(
>             pred_vertices[:, list_id, :],
>             rotation=torch.eye(3, device=device).unsqueeze(0).expand(
>                 batch_size, -1, -1),
>             translation=camera_translation,
>             focal_length=constants.FOCAL_LENGTH,
>             camera_center=camera_center,
>         )
>         # [-h/2, h/2] + h/2 -> [0, h] shape [1,1,2]
>         outline_points_2d[:, :, 0] = outline_points_2d[:, :, 0] + width / 2
>         outline_points_2d[:, :, 1] = outline_points_2d[:, :, 1] + height / 2
> ```

Hi, have you visualized the 

> I tried it.
> 
> 1. At 217th line of `up/render.py`, it does render.
>    `renderings = render(model['neutral'], image, cam, steps, segmented, scale)`.
> 2. At 156th line of `up/render.py` in `render` function, it creates renderer.
> 
> ```
>     rn = _create_renderer(w=int(w * scale),
>                           h=int(h * scale),
>                           near=4.,
>                           far=30.+dist,
>                           rt=np.array(cam['rt']),
>                           t=np.array(cam['t']),
>                           f=np.array([cam['f'], cam['f']]) * scale,
>                           # c=np.array(cam['cam_c']),
>                           texture=texture)
> ```
> 
> 2-A. At 63th line of `up/render.py` in `_create_renderer` function, it projects SMPL vertices to image plane.
> `rn.camera = ProjectPoints(rt=rt, t=t, f=f, c=c, k=k)`
> 
> 1. At 170th line of `up/render.py` in `render` function, it iteratively render mesh from different view.
> 
> ```
>         imtmp = _simple_renderer(rn=rn,
>                                  meshes=[mesh],
>                                  yrot=light_yrot,
>                                  texture=texture)
> ```
> 
> 1. At 131th line of `up/render.py` in `_simple_renderer` function, it returns `rn.r`.
>    `return rn.r`. I need to know what is `r`.
> 2. At 81th line of [`opendr/camera.py`](https://github.com/mattloper/opendr/blob/master/opendr/camera.py), in `compute_r` function of `ProjectPoints` class, it returns `self.r_and_derivatives[0].squeeze()`.
> 3. At 143th line of `opendr/camera.py` in `r_and_derivatives` function, it calls `cv2.projectPoints`. Here, there are five inputs of this function. `v, self.rt.r, self.t.r, self.camera_mtx, self.k.r`. `rt` and `t` are defined in 156th line of `up/render.py`.
> 
> ```
> self.rt.r: np.array(cam['rt'])
> self.t.r: t=np.array(cam['t'])
> ```
> 
> `k` is defined in 56th line of 'up/render.py' as `k = np.zeros(5)`.
> 
> `camera_mtx` is obtained from 134th line of `opendr/camera.py`, and `f` is from 156th line of `up/render.py` `c` is from 55th line of `up/render.py`.
> `f=np.array([cam['f'], cam['f']]) * scale,`
> `c = np.array([w, h]) / 2.`
> where `w` and `h` are image resolution from 214th line of `up/render.py`.
> 
> The result image looks like below. Green dots are `j2d` and blue dots are my projected image points.
> ![2](https://user-images.githubusercontent.com/13601723/68518340-d77c7d80-02ce-11ea-96a3-7a8d5d01fb03.jpg)
> 
> When I use `c = np.array([crop_info[1], crop_info[0]]) / 2.` like `pose/tools/create_dataset.py`, the result image looks like below. Green dots are `j2d` and blue dots are my projected image points. As I defined `c` using `crop_info`, I cropped image using `crop_info[2], [3], [4], [5]`and resized it to `crop_info[0], [1]`. For example, `'img = cv2.imread(filename)`, `cropped_img = img[crop_info[2]:crop_info[3], crop_info[4]:crop_info[5],:].`, `resized_img = cv2.resize(cropped_img, (crop_info[1], crop_info[0]))`. Green dots and blue dots are visualized on `resized_img`.
> ![2](https://user-images.githubusercontent.com/13601723/68518377-0692ef00-02cf-11ea-9943-070a369adf98.jpg)
> 
> Note that I used `cv2.projectPoints` function like opendr. Could you tell me what I've done wrong? and why the result image looks strange when I use `c = np.array([w, h]) / 2.`?

I explored the dataset and found a constant 20px offset could be applied to your method. Hope that helps!
